src/enumerable.cr

```diff
@@ -984,6 +1019,43 @@ module Enumerable(T)
     max_by? &.itself
   end
 
+  # Returns an array of the maximum `k` elements, sorted descending.
+  #
+  # It compares using `<=>` so it will work for any type that supports that method.
+  #
+  # ```
+  # [7, 5, 2, 4, 9].max(3) # => [9, 7, 5]
+  # ["Eve", "Alice", "Bob", "Mallory", "Carol"].max(2) # => ["Mallory", "Eve"]
+  # ```
+  #
+  # Raises `Enumerable::ArgumentError` if `k` is negative or if any elements are
+  # not comparable.
+  def max(k : Int) : Array(T)
+    raise ArgumentError.new("negative size #{k}") if k < 0
+    data = to_a
+    n = data.size
+    k = n if k > n
+    (0..k-1).map do |i|
+      r = qselect_internal(data, 0, n-1, n-1-i)
+      raise ArgumentError.new("comparison failed") if r.nil?
+      r
+    end
+  end
+
+  # Like `max(k)` but returns `nil` if `k` is negative or if any elements are not
+  # comparable.
```

Comment:
Failure to compare two elements should raise, even for the nilable methods. This is the convention we adopt in `Enumerable`.